src/main/java/software/amazon/smithy/lsp/ext/Completions.java

```diff
@@ -41,24 +62,56 @@ private Completions() {
      *
      * @param model Smithy model
      * @param token token
+     * @param isTrait boolean
+     * @param target Optional ShapeId of the target trait target
      * @return list of completion items
      */
-    public static List<SmithyCompletionItem> find(Model model, String token) {
+    public static List<SmithyCompletionItem> find(Model model, String token, boolean isTrait,
+                                                  Optional<ShapeId> target) {
         Map<String, SmithyCompletionItem> comps = new HashMap<>();
         String lcase = token.toLowerCase();
 
+        Set<ShapeId> shapeIdSet;
+        // If the token is part of a trait statement, filter the set to trait shapes which can be applied to the shape
+        // that the trait targets.
+        if (isTrait) {
+            shapeIdSet = model.shapes()
+                    .filter(shape -> shape.hasTrait(ShapeId.from("smithy.api#trait")))
+                    .filter(shape -> {
+                        if (!target.isPresent()) {
+                            return true;
+                        }
+                        return shape.getTrait(TraitDefinition.class).get().getSelector().shapes(model)
```

Comment:
Can use `expectTrait` since it was filtered before.

Reply:
Addressed in 43489272c27970304970f2d2e616c95c47cc1c00.